backend/services/notificationService.ts

```diff
@@ -1,60 +1,102 @@
-    return this.notifyApi.sendEmail(templateId, email, {
+  private async sendEmail({ templateId, emailAddress, personalisation }): Promise<void> {
+    return this.notifyApi.sendEmail(templateId, emailAddress, {
       personalisation,
       reference: null,
     })
   }
 
-  private async getUserDetails(context: Context, username: string) {
+  private async getUserDetails(context: Context, username: string): Promise<UserDetails> {
     const [{ email }, { name }] = await Promise.all([
       this.oauthApi.userEmail(context, username),
       this.oauthApi.userDetails(context, username),
     ])
     return { email, name }
   }
 
-  private getEmailAddress(recipient: Recipient, agencyId: string, userEmail: string): string {
+  private async getEmailAddress(
+    context: Context,
+    recipient: Recipient,
+    agencyId: string,
+    userEmail: string
+  ): Promise<string> {
     switch (recipient) {
       case 'user':
         return userEmail
       case 'omu':
-        return notifications.emails[agencyId]?.omu
+        return this.prisonRegisterApi.getOffenderManagementUnitEmailAddress(context, agencyId)
       case 'vlb':
-        return notifications.emails[agencyId]?.vlb
+        return this.prisonRegisterApi.getVideoLinkConferencingCentreEmailAddress(context, agencyId)
       default:
         throw new Error(`could not send email to ${recipient}`)
     }
   }
 
-  private async sendEmails(context: Context, username: string, spec: EmailSpec): Promise<void> {
-    const { email: userEmail, name: usersName } = await this.getUserDetails(context, username)
+  private async sendEmails(context: Context, username: string, emailSpec: EmailSpec): Promise<void> {
+    const userDetails = await this.getUserDetails(context, username)
+    const recipientEmailer = recipientEmailSpec =>
+      this.emailToRecipient(context, recipientEmailSpec, userDetails, emailSpec)
+    await Promise.all(emailSpec.recipients.map(recipientEmailer))
+  }
+
+  private async emailToRecipient(
+    context: Context,
+    recipientEmailSpec: RecipientEmailSpec,
+    userDetails: UserDetails,
+    emailSpec: EmailSpec
+  ): Promise<void> {
+    try {
+      const emailAddress = await this.getEmailAddress(
+        context,
+        recipientEmailSpec.recipient,
+        emailSpec.agencyId,
+        userDetails.email
+      )
 
-    spec.recipients.forEach(email => {
-      const emailAddress = this.getEmailAddress(email.recipient, spec.agencyId, userEmail)
-      if (emailAddress) {
-        this.sendEmail({
-          templateId: email.template,
-          email: emailAddress,
-          personalisation: email.personalisation(usersName),
-        }).catch(error => {
-          log.error(
-            `Failed to email the ${email.recipient} a ${spec.name}: ${error.message}`,
-            error.response?.data?.errors
-          )
-        })
+      if (!emailAddress) {
+        return Promise.resolve()
       }
-    })
+
+      return this.sendEmail({
```

Comment:
not sure this is quite the same behaviour as before - it was quite fire and forget before whereas this looks like it will propagate errors to the client?

Reply:
I changed the sendEmails method so that it waited for all the sendEmail calls to resolve before returning (using Promise.all at line 67.  This makes notificationService.ts easier to test.  The alternative would require using a timer in order to to wait for the fire and forget to complete before making assertions.  This change was required because fetching the e-mail addresses is now an async operation.

I changed the client (bookingService.ts) so that that the public methods of notificationService are called without an await.  In other words sending e-mails is still fire-and-forget, but the place where that happens has been moved from notification service to booking service.  I did this because it makes testing simpler.

Reply:
Also changed log level from `warn` to `info` for 404 HTTP status in ouathEnabledClient.js